resourcemanager/pooltask/task.go

```diff
@@ -127,35 +129,40 @@ type GPool[T any, U any, C any, CT any, TF Context[CT]] interface {
 
 // TaskController is a controller that can control or watch the pool.
 type TaskController[T any, U any, C any, CT any, TF Context[CT]] struct {
-	pool     GPool[T, U, C, CT, TF]
-	close    chan struct{}
-	wg       *sync.WaitGroup
-	taskID   uint64
-	resultCh chan U
+	pool           GPool[T, U, C, CT, TF]
+	productCloseCh chan struct{}
+	wg             *sync.WaitGroup
+	prodWg         *sync.WaitGroup
+	taskID         uint64
+	resultCh       chan U
 }
 
 // NewTaskController create a controller to deal with pooltask's status.
-func NewTaskController[T any, U any, C any, CT any, TF Context[CT]](p GPool[T, U, C, CT, TF], taskID uint64, closeCh chan struct{}, wg *sync.WaitGroup, resultCh chan U) TaskController[T, U, C, CT, TF] {
+func NewTaskController[T any, U any, C any, CT any, TF Context[CT]](p GPool[T, U, C, CT, TF], taskID uint64, productCloseCh chan struct{}, wg, prodWg *sync.WaitGroup, resultCh chan U) TaskController[T, U, C, CT, TF] {
 	return TaskController[T, U, C, CT, TF]{
-		pool:     p,
-		taskID:   taskID,
-		close:    closeCh,
-		wg:       wg,
-		resultCh: resultCh,
+		pool:           p,
+		taskID:         taskID,
+		productCloseCh: productCloseCh,
+		wg:             wg,
+		prodWg:         prodWg,
+		resultCh:       resultCh,
 	}
 }
 
 // Wait is to wait the pool task to stop.
 func (t *TaskController[T, U, C, CT, TF]) Wait() {
-	<-t.close
+	t.prodWg.Wait()
 	t.wg.Wait()
 	close(t.resultCh)
 	t.pool.DeleteTask(t.taskID)
 }
 
 // Stop is to send stop command to the task. But you still need to wait the task to stop.
 func (t *TaskController[T, U, C, CT, TF]) Stop() {
+	close(t.productCloseCh)
+	t.prodWg.Wait()
 	t.pool.StopTask(t.TaskID())
+	channel.Clear(t.resultCh)
```

Comment:
line162 close `productCloseCh`, but it is no consumer consumption task. Is it a deadlock？

Reply:
Fix